dns/ddclient/src/opnsense/mvc/app/models/OPNsense/DynDNS/DynDNS.xml

```diff
@@ -110,6 +110,14 @@
                     <Required>N</Required>
                     <mask>/^[^\n]*$/</mask>
                 </password>
+                <netcupAPIKey type="UpdateOnlyTextField">
+                    <Required>N</Required>
+                    <mask>/^[^\n]*$/</mask>
+                </netcupAPIKey>
+                <netcupSetTTL type="BooleanField">
```

Comment:
TTL will be added in a more generic way in https://github.com/opnsense/plugins/pull/3550/files#diff-ba4d24cee94fcdc665a4223826a1daea10fde3e20971b1e4865ee261d8e2b5f4R174

we should try to avoid service-specific naming

Reply:
TTL is in master now

Reply:
I have rebased my code against latest master and made the following changes:
- all Netcup related fields were removed. The plugin now only uses existing fields.
- some polishing (removed unused exports, better logging and parameter checking)

Reply:
It would probably make more sense then to use user+pass as API key+pass and move customer to resourceID. But I think it's required and the GUI will not require it in this case? It needs some sort of constraint added but we can do that post-merge.

Reply:
Uh, that would be even more confusing, IMHO.
There is a service specific field for the azure plugin, so why can't another plugin have it's own specific field, too? Yes, with more plugins coming up this can clutter the configuration, I agree.
But abusing an already very specific field for yet another specific purpose is IMHO not a good way forward.

I still believe that every service should have it's very own custom fields with an exact help of what they are for. That would be the best option. But I guess that is not possible with the way the framework works, right?

In that case let me propose another way:
Get rid of all service specific fields and introduce a couple of general purpose fields. If a service then needs some special configuration, it can use those custom fields. These would then all be documented in the Wiki.

Does that sound like a way forward?

Reply:
> I still believe that every service should have it's very own custom fields with an exact help of what they are for. That would be the best option. But I guess that is not possible with the way the framework works, right?

Rather not as in practice 99% of the terminology is the same, in this case netcup uses a `key`, `password` and a customer number, in which case the customer number is the only outlier here.  The idea to use resourceID as customer in this case isn't very weird, but if we can think of a more generic term, sure. Just remember data normalization exist for a reason....

For plugins we don't maintain (or support) at all, we don't care that much, in case you are going to ask about the `acme-client` plugin. The lack of normalization always lead to technical debt at some point in time.

>Get rid of all service specific fields and introduce a couple of general purpose fields. If a service then needs some special configuration, it can use those custom fields. These would then all be documented in the Wiki.

That is already the case, based on what most of these services require.

Reply:
Yeah, I don’t care which way, but no specific fields. Sounds like a customer field can be introduced, but let’s not call it „customer ID“ because that is what this provider is using. User and pass are then the API stuff.

Reply:
Guys, I think you are talking past each other.

Netcup requires as login credentials something like a user name and a password: User name is the customer number and password is the so called API password. **You want Curly 060 to use no custom fields for this? This is exactly what Curly060 has implemented:** Instead if introducing custom fields he used the standard fields for username and password for this.

BUT: Netcup **requires an additional secrect** for accessing the DDNS API, and this is the field called "API key". The API key is a secret and **MUST** be hidden in the GUI. **There is no existing standard field available**, that's why Curly introduced a new one and made it invisible for all other DDNS plugins. Exactly like the Azure plugin does this with the resourceID. Your proposal to use resourceID is not feasible, because resourceID neither does not match "API Key" semantically nor is a field protected from view.

I recommend to follow Curlys implementation and add a new field for the key. **If we need it generically, we could call it "Key" instaed of "Netcup API Key".**

Reply:
I recommend to follow the review comments. We don’t need every provider to reinvent the wheel and have the rest of the world deal with the implications. You can always try to write a separate plugin for the provider with all the upsides and downsides.

Reply:
> I recommend to follow the review comments. 
Which one exactly?

You proposed to introduce a customer field:
>Sounds like a customer field can be introduced
There is no need for a customer field, Curly060 used the username for this. It's the API key which needs a field.

Reply:
What’s a key and what’s a secret and what’s a password? What’s a user and what’s a customer. Choosing three things out of two is mildly annoying. Ideally you’d want to condense the API stuff into a single token and be done with it. Now you have a user and password. That‘s also what he suggested but then we got stuck over service-specific fields which we will have to rigorously vet.

Reply:
I can't follow. You might know that it's common that some Web APIs require pairs of secrets. I have the feeling you are not really reading what others are writing. Please explain: Why is it okay to introduce a second visible field "resourceID" instead of concatenating it's value into the user ID field, but it's not okay to introduce a second invisible field "key" or "client secret"? Can you please come up with an logical explaination for that?

Reply:
Let’s try to keep this between author and reviewer. It’s hard enough to follow as is. 😉

Reply:
Curly060, can I be become Co-Author? Otherwise I'm not accept here.

Every DDNS has the same problem:
1. Determine the As-Is value (= quering the master DDNS server)
2. Determine and monitor the To-Be value
3. Update the value if needed

Only step no 3 is different for different providers. For that reason it is a good idea to have one OPNSense plugin for all DDNS tasks and sub plugins for the different provider. The proposal to come up with an own OPNSense plugin for Netcup is for sure an antipattern. The discussion for a solution for the missing secret field is blocked. So I cannot see a possible innovation progress here.

Reply:
Dude, nothing is blocked. Two proposals are open for implementation. If you think the suggestion to create a separate plugin is overkill then I don’t disagree. I was merely pointing out when it makes sense to use „netcup“ specific fields and when it doesn’t.

Reply:
Let's come to an end with this. I have now added a field called "clientSecret" of type "password" with label text "Additional client secret, e.g. an API key". Now the field has nothing to do with Netcup at all, except that the Netcup plugin is currently its only user. But since the name is quite generic, this might change with future plugins coming up. Plus there are already fields that are provider specific and have generic names (ttl, resourceId). So I think this is a compromise between usability and maintainability.

Is the current form acceptable now?